exploration/selection-declaration.md

```diff
@@ -149,14 +150,11 @@ _What prior decisions and existing conditions limit the possible design?_
 
 ## Proposed Design
 
-_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+The design alternative [Match on variables instead of expressions](#match-on-variables-instead-of-expressions)
```

Comment:
There is also PR #867

I suggested an option there:
https://github.com/unicode-org/message-format-wg/pull/867#discussion_r1729318442

But we discuss the same thing in so many places that I don't know where to comment anymore...
Let me know if you want me to copy it here.

Reply:
@mihnita I created #867 to create a quick update to the design doc to reflect the discussion that came up in [Monday's 2024-08-19 meeting](https://github.com/unicode-org/message-format-wg/blob/main/meetings/2024/notes-2024-08-19.md), as @aphillips's suggested then, and because I couldn't make them as suggested changes in this PR since it's outside the bounds of Github's diff view. I think a separate PR is the way to do it.

(Btw, regarding the idea itself from [your comment](https://github.com/unicode-org/message-format-wg/pull/867#discussion_r1729318442), I agree with [@bearfriend's comment](https://github.com/unicode-org/message-format-wg/pull/867/files#r1729388338) in response. I have more explanation to offer to that position. I can share that with you in a 1:1 chat in case that would affect whether you want to pursue a small PR that adds the option to the design doc so that it gets equal consideration.)

Reply:
That discussion is not very clear. Can you boil it down to a reasonable use case where the selection should differ from the formatting semantics (eg disregarding different number systems).

Reply:
Isn't that covered by the added example?
```
.match {$person :gender}
male {{Bienvenido {$person :personName}}}
female {{Bienvenida {$person :personName}}}
other {{Le damos la bienvenida {$person :personName}}}
```
One can also imagine similar patterns for non-complex inputs such as `{$time :partOfDay}` (morning/afternoon/evening/night), `{$adjective :affect}`, or really any kind of "binning" classification into finite categories for distinct treatment that still needs the original value.

Reply:
I find the "Do Nothing" alternative problematic because it does not signal anything wrong with a message like
```
.match {$count :integer}
one {{You have {$count} whole apple.}}
* {{You have {$count} whole apples.}}
```

If that message is formatted with `{ count: 1.2 }`, the result is "You have 1.2 whole apple." which is clearly not what was intended by its author.

This problem is the main reason that I filed #736 originally, and which I think we need some solution for. As we've discussed on many occasions, MF2 will always be an auxiliary language to its users, and we should not expect users to always remember its details. Therefore, I think it's important that we find some way for the above message to be considered invalid, so that this mistake is caught as early as possible, or we add a side effect to `.match` so that the desired behaviour is realised.

My own first preference is the alternative proposed in this PR, but I've also filed #860 to provide a next-best choice that retains our current syntax, but signals an error in this particular case, where a selector variable is used in a placeholder.

Reply:
Thinking about this some more. My proposal is:
- in a match statement, options effectively mutate variables (see below) 
- it is illegal for a variable to occur twice in a match statement.

This keeps most message more concise, producing the expected results: eg without the problems that occur above with input of count : 1.2

Example1.

```
.match {$count :integer}
one {{You have {$count} whole apple.}}
* {{You have {$count} whole apples.}}
```

is precisely equivalent to:

```
.local $count2 {$count :integer}
.match {$count2}
one {{You have {$count2} whole apple.}}
* {{You have {$count2} whole apples.}}
```

Example 2.
And the following is illegal.
```
.match {$count <anything>}{$count <anything>}
```

 The message author is required to **rewrite** it explicitly, eg to:
```
.local $count3 {$count}
.match {$count <anything1>}{$count3 <anything2>}
```
I really don't think this is a burden, since the number of times the same variable is used twice in a match (or the older Select) is vanishingly small. Since it is an error — and the advice to fix is easy — that will prevent misbehavior.

Reply:
@macchiati 
> I really don't think this is a burden, since the number of times the same variable is used twice in a match (or the older Select) is vanishingly small. 

This is true of numbers. Dates offer an alternative... or person names in which one might want multiple selection from a complex value.

Note: I'm not arguing against the proposal.